### DISCUSSION TITLE
Modbus map - potentially allowing other inverter platforms

### DISCUSSION THREAD

**Discussion by n2aws** [binsentsu/home-assistant-solaredge-modbus#202] [Ideas]
Hello,

I use home assistant and was getting ready to create an integration for my platform, when I came across your project.

Rather than reinvent the wheel, I was wondering if you'd be interested in changing the modbus definitions slightly.

Rather than hardcoding addresses, we could define them (similar to how translations are done now) which would allow this project to be used across any platform that someone creates a modbus map for.

Obviously this would need the ability to define things like how many inverters there are, and maybe a UI component to select which inverter platform "template" to use and similar tasks..

As an example,  I have a schnieder based system.   Almost all of the data you pull for this project, is also available in the Schneider ecosystem (at different modbus addresses and device IDs),  so I'm envisioning something like a UI dropdwn selectable to switch between "SolarEdge" and "Schneider",  and as a result of that selection, it pulls in the appropriate modbus addresses from a "library" we create.  I tell it the IP (or which USB port) to use and the device ID(s),  and voila.. it reports the exact same data you currently do.

I also use Tigo optimizers, and could add a modbus map for that system too,  although it'd only be available via a physical connection.

And finally, I have EG4 batteries, and can add a modbus map for those as well,  though that is a bigger conversation (monitoring charge controllers and battery states)

Barring any interest in that,  would you be opposed to my forking the code, and making a "Schneider" specific implementation, where 95% of the code would be identical?

**Comment by binsentsu** [comment]
Hi,
Thanks for the interest, but I don't think that this integration is the correct base to start from. When looking broader, all these inverters share the same underlying sunspec. It seems that [this repo](https://github.com/CJNE/ha-sunspec) already tries to generify in a way which I think you would do it. The problem is also that mostly all these manufacturers give some twists to the spec implementation and then you end up with some quirky conditions in the codebase. Furthermore this integration is far from platinum in the HA quality scale, so mayb a fresh start, taking code from this repo into consideration, would be more benifical in the long run.
